### PR TITLE
1k+ assets (part 5)

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

1. Extend `apply.Balances` interface with `GetEx` methods. 
2. Use `Allocate`/`Deallocate` for asset holdings adding/removing.
3. Implement ledger.LookupFull that loads all the assets.

### CODE REVIEW

File: ledger/cow.go [hunk 1]
```diff
@@ -126,15 +128,48 @@ func (cb *roundCowState) getCreator(cidx basics.CreatableIndex, ctype basics.Cre
...
+	// data from cb.mods.Accts is newer than from lookupParent.lookupHolding, so add the asset if any
+	if holding, ok := d.AccountData.Assets[basics.AssetIndex(cidx)]; ok {
+		ad.Assets[basics.AssetIndex(cidx)] = holding
+	}
+	return ledgercore.PersistedAccountData{AccountData: ad}, nil
+
+	// if err != nil {
+	// 	// save ExtendedHolding portion if possible for later reference
+	// 	cb.mods.Accts.Register(addr, d)
+	// }
```
Comment (algorandskiy) [inline ledger/cow.go:170]: removed in part-6

File: ledger/appcow.go [hunk 2]
```diff
@@ -200,53 +200,6 @@ func errAlreadyStorage(addr basics.Address, aidx basics.AppIndex, global bool) e
 	return fmt.Errorf("%v has already opted in to app %d", addr, aidx)
 }
 
-// Allocate creates kv storage for a given {addr, aidx, global}. It is called on app creation (global) or opting in (local)
-func (cb *roundCowState) Allocate(addr basics.Address, aidx basics.AppIndex, global bool, space basics.StateSchema) error {
-	// Check that account is not already opted in
-	allocated, err := cb.allocated(addr, aidx, global)
-	if err != nil {
-		return err
-	}
```
Comment (algorandskiy) [inline ledger/appcow.go:209]: moved to **creatablecow.go** since where two methods for both apps and assets

File: ledger/cow.go [hunk 3]
```diff
@@ -126,15 +128,48 @@ func (cb *roundCowState) getCreator(cidx basics.CreatableIndex, ctype basics.Cre
...
 	if ok {
-		return d, nil
+		return ledgercore.PersistedAccountData{AccountData: d}, nil
 	}
 
 	return cb.lookupParent.lookup(addr)
 }
 
+// lookupWithHolding is gets account data but also fetches asset holding or app local data for a specified creatable
+func (cb *roundCowState) lookupHolding(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType) (data ledgercore.PersistedAccountData, err error) {
```
Comment (algorandskiy) [inline ledger/cow.go:141]: use `cidx basics.CreatableIndex` for possible generalization in future

File: ledger/acctupdates.go [hunk 4]
```diff
@@ -1612,6 +1631,86 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
 	}
 }
 
+// lookupFullWithRewards returns the full account data for a given address at a given round.
+// The rewards are added to the AccountData before returning. Note that the function doesn't update the account with the rewards,
+// even while it does return the AccoutData which represent the "rewarded" account data.
+func (au *accountUpdates) lookupFullWithRewards(rnd basics.Round, addr basics.Address) (pad ledgercore.PersistedAccountData, err error) {
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1637]: this would not work correctly - you need to make all the reads from the database within the same `RLock`. Taking the lock twice means that in-between there could be a database write.

File: ledger/apply/apply.go [hunk 5]
```diff
@@ -31,6 +31,9 @@ type Balances interface {
 	// A non-nil error means the lookup is impossible (e.g., if the database doesn't have necessary state anymore)
 	Get(addr basics.Address, withPendingRewards bool) (basics.AccountData, error)
 
+	// GetEx is like Get(addr, false), but also loads specific creatable
+	GetEx(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType) (basics.AccountData, error)
```
Comment (tsachiherman) [inline ledger/apply/apply.go:35]: I think that a single `Get` method would be preferable, that would have the following parameters:
Get(addr basics.Address, withPendingRewards bool, cidx basics.CreatableIndex, ctype basics.CreatableType)
where we ignore cidx of -1, and adding support for ctype of "AssetParams" or something like that.
(i.e. so that this Get call would be good for asset holding/asset params and applications )

File: ledger/apply/asset.go [hunk 6]
```diff
@@ -102,7 +103,14 @@ func AssetConfig(cc transactions.AssetConfigTxnFields, header transactions.Heade
...
 
-		return balances.PutWithCreatable(header.Sender, record, created, nil)
+		// Save asset params and holding, and register new creatable
+		err = balances.PutWithCreatable(header.Sender, record, created, nil)
+		if err != nil {
+			return err
+		}
+
+		// register new holding
+		return balances.Allocate(header.Sender, basics.CreatableIndex(newidx), basics.AssetCreatable, false, basics.StateSchema{})
```
Comment (tsachiherman) [inline ledger/apply/asset.go:113]: maybe: pass the `created` to `Allocate`, instead of passing all the arguments ?

File: ledger/apply/asset.go [hunk 7]
```diff
@@ -140,6 +148,8 @@ func AssetConfig(cc transactions.AssetConfigTxnFields, header transactions.Heade
 
 		delete(record.Assets, cc.ConfigAsset)
 		delete(record.AssetParams, cc.ConfigAsset)
+
+		balances.Deallocate(creator, basics.CreatableIndex(cc.ConfigAsset), basics.AssetCreatable, false)
```
Comment (tsachiherman) [inline ledger/apply/asset.go:152]: I'm not a big fan of the usage of the `basics.CreatableLocator` above, since it promotes the `missed field initialization` type-bug. I'd rather seeing the object being constructed as a result of `makeCreatableLocator(..)`

File: ledger/cow.go [hunk 8]
```diff
@@ -126,15 +128,43 @@ func (cb *roundCowState) getCreator(cidx basics.CreatableIndex, ctype basics.Cre
...
+		} else {
+			_, exist = ad.AppLocalStates[basics.AppIndex(cidx)]
+		}
+
+		if exist {
+			return ledgercore.PersistedAccountData{AccountData: ad}, nil
+		}
+	}
+
+	d, err := cb.lookupParent.lookupHolding(addr, cidx, ctype)
```
Comment (tsachiherman) [inline ledger/cow.go:156]: is this always needed ? i.e. in case the `ad` already contain the holding, there is no need to lookup, right ?

File: ledger/ledgercore/statedelta.go [hunk 9]
```diff
@@ -70,12 +70,24 @@ type StateDelta struct {
...
+)
+
 // AccountDeltas stores ordered accounts and allows fast lookup by address
 type AccountDeltas struct {
 	// actual data
 	accts []basics.BalanceRecord
 	// cache for addr to deltas index resolution
 	acctsCache map[basics.Address]int
+	// holdings keeps track of created and deleted holdings per address
+	holdings map[basics.Address]map[basics.AssetIndex]HoldingAction
```
Comment (tsachiherman) [inline ledger/ledgercore/statedelta.go:90]: I can't see here were you're using this `holding` structure, but it looks here as we have some first-level duplication here : the data inside this map should go into the `accts`.

**Review by tsachiherman — CHANGES_REQUESTED** [review]
it looks like series of steps in the right direction, although it would require few more tweaks. ( I'm mostly concerned about the database locks, but there are few other points as well )

File: ledger/acctupdates.go (hunk 4 above)
Comment (algorandskiy) [reply ledger/acctupdates.go:1637]: thanks, fill fix here and below.

File: ledger/apply/apply.go (hunk 5 above)
Comment (algorandskiy) [reply ledger/apply/apply.go:35]: I want to emphasize the fact `Ex` method is more expensive.

File: ledger/apply/asset.go (hunk 6 above)
Comment (algorandskiy) [reply ledger/apply/asset.go:113]: I'll check if it suitable for apps as well

File: ledger/apply/asset.go (hunk 7 above)
Comment (algorandskiy) [reply ledger/apply/asset.go:152]: could you elaborate please? not sure I follow the point of using `makeCreatableLocator` here

File: ledger/cow.go (hunk 8 above)
Comment (algorandskiy) [reply ledger/cow.go:156]: yes, if ad in `mods` and the asset exists, it is returned above

File: ledger/ledgercore/statedelta.go (hunk 9 above)
Comment (algorandskiy) [reply ledger/ledgercore/statedelta.go:90]: this map tracks created and **deleted** holdings. See `SetHoldingDelta` below.

File: ledger/ledgercore/statedelta.go [hunk 10]
```diff
@@ -70,12 +70,24 @@ type StateDelta struct {
 	PrevTimestamp int64
 }
 
+// HoldingAction is an enum of actions on holdings
+type HoldingAction uint64
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/ledgercore/statedelta.go:74]: will do

**Comment by algorandskiy** [comment]
Rebased, requires #2069 to be merged first.

**Comment by algorandskiy** [comment]
Rebased again.

File: ledger/accountdb.go [hunk 11]
```diff
@@ -900,31 +933,86 @@ func (qs *accountsDbQueries) lookupCreator(cidx basics.CreatableIndex, ctype bas
...
+		// Return the zero value of data
+		return data, fmt.Errorf("unable to query account data for address %v : %w", addr, err)
+	}
+
+	return
+}
+
+func lookupExt(rdb db.Accessor, addr basics.Address, extension func(*sql.Stmt, *ledgercore.PersistedAccountData) error) (data dbAccountData, err error) {
+	rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		lookupStmt, err := tx.Prepare(lookupAcctBaseQuery)
```
Comment (tsachiherman) [inline ledger/accountdb.go:969]: don't forget to `Close` the prepared statement once you no longer need it.

File: ledger/acctupdates.go [hunk 12]
```diff
@@ -1794,7 +1814,10 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
 			// Check if this is the most recent round, in which case, we can
 			// use a cache of the most recent account state.
 			if offset == uint64(len(au.deltas)) {
-				return macct.data, nil
+				if full && macct.data.ExtendedAssetHolding.Count != 0 {
+					macct.data.Assets, macct.data.ExtendedAssetHolding, err = au.accountsq.loadHoldings(macct.data.ExtendedAssetHolding)
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1818]: in order to make sure that the return is valid, you need to join it with the round number and compare it with `currentDbRound`. Otherwise, you might have just created a mixed-up balance record of old account data with new extended holding.

File: ledger/acctupdates.go [hunk 13]
```diff
@@ -1911,7 +1992,16 @@ func (au *accountUpdates) lookupWithoutRewards(rnd basics.Round, addr basics.Add
 		// present in the on-disk DB.  As an optimization, we avoid creating
 		// a separate transaction here, and directly use a prepared SQL query
 		// against the database.
-		dbad, err = au.accountsq.lookup(addr)
+
+		if extension != nil {
+			dbad, err = lookupExt(au.dbs.Rdb, addr, extension)
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1997]: it's abit hard to follow, but I'm not sure if the extension method here is called from the *same* transaction or not. It's vital to call it from the same read transaction - otherwise, the `dbad.round` comparison below would not be sufficient.

File: ledger/apply/asset.go (hunk 7 above)
Comment (tsachiherman) [reply ledger/apply/asset.go:152]: sure;
How can you tell that you did not forget to initialize one of the fields in the `deleted` object ?
Creating a wrapper function around that would ensure that you won't be forgetting to initialize all the field members, even when you add a new field to the structure.

File: ledger/ledgercore/statedelta.go (hunk 9 above)
Comment (tsachiherman) [reply ledger/ledgercore/statedelta.go:103]: yes, I realize that.. I'm wondering why wound't we add this map to the `PersistedBalanceRecord` instead ?

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Few open questions above; one particular issue is the missing atomicity verification on database reads.

File: ledger/acctupdates.go (hunk 13 above)
Comment (algorandskiy) [reply ledger/acctupdates.go:1997]: not sure I understand. `au.accountsq.lookup` uses prepared queries from `au.dbs.Rdb`, same way as `lookupExt` does. `lookupExt` prepares and runs two queries on the same transaction derived from `au.dbs.Rdb.Atomic`. It looks like as a complete equivalent of the old code.

File: ledger/accountdb.go [hunk 14]
```diff
@@ -900,31 +933,86 @@ func (qs *accountsDbQueries) lookupCreator(cidx basics.CreatableIndex, ctype bas
...
+	if err == sql.ErrNoRows {
+		// Return the zero value of data
+		return data, fmt.Errorf("unable to query account data for address %v : %w", addr, err)
+	}
+
+	return
+}
+
+func lookupExt(rdb db.Accessor, addr basics.Address, extension func(*sql.Stmt, *ledgercore.PersistedAccountData) error) (data dbAccountData, err error) {
+	rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
```
Comment (tsachiherman) [inline ledger/accountdb.go:968]: You'll need to add `err = rdb.Atomic` in order to catch `Commit` errors as well.

Comment (tsachiherman) [reply ledger/accountdb.go:968]: Also, I think that you'd like to use `db.Retry` here.

File: ledger/acctupdates.go (hunk 13 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1997]: ah - I missed that. Thanks.

File: ledger/apply/asset.go (hunk 7 above)
Comment (algorandskiy) [reply ledger/apply/asset.go:152]: Ok, I'll submit a separate PR to master with this fix since it is not related to this PR.

File: ledger/ledgercore/statedelta.go (hunk 9 above)
Comment (algorandskiy) [reply ledger/ledgercore/statedelta.go:103]: It is possible but `PersistedBalanceRecord` is also used in eval for accounts prefetching and in that context it does not need holdings delta at all. I'll revisit your suggestion as move forward with params implementation.

File: ledger/acctupdates.go [hunk 15]
```diff
@@ -1856,9 +1888,18 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
...
+				goto retryAcquireLock
+			}
 		} else {
 			dbad, err = au.accountsq.lookup(addr)
 		}
+
+		if err != nil {
+			return ledgercore.PersistedAccountData{}, err
+		}
+
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1902]: not sure we miss this error check in master, added

File: ledger/acctupdates.go (hunk 12 above)
Comment (algorandskiy) [reply ledger/acctupdates.go]: done

File: ledger/acctupdates.go [hunk 16]
```diff
@@ -1825,26 +1894,88 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
 		// present in the on-disk DB.  As an optimization, we avoid creating
 		// a separate transaction here, and directly use a prepared SQL query
 		// against the database.
-		dbad, err = au.accountsq.lookup(addr)
+		if full {
+			dbad, err = lookupFull(au.dbs.Rdb, addr)
+			var mismatchRoundErr *MismatchingDatabaseRoundError
+			if errors.As(err, &mismatchRoundErr) {
+				goto retryAcquireLock
+			}
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1902]: MismatchingDatabaseRoundError might happen when round advances between accountbase and accountext reading, not sure to to test it.

File: ledger/acctupdates.go [hunk 17]
```diff
@@ -1814,7 +1874,16 @@ func (au *accountUpdates) lookupWithRewards(rnd basics.Round, addr basics.Addres
...
 			au.baseAccounts.writePending(macct)
-			return macct.pad.AccountData, nil
+			if full && macct.pad.ExtendedAssetHolding.Count != 0 {
+				macct.pad.Assets, macct.pad.ExtendedAssetHolding, err = au.accountsq.loadHoldings(macct.pad.ExtendedAssetHolding, macct.round)
+				var mismatchRoundErr *MismatchingDatabaseRoundError
+				if errors.As(err, &mismatchRoundErr) {
+					// go to waiting
+					// note, au.accountsMu is still held
+					goto retryLocked
+				}
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1884]: MismatchingDatabaseRoundError might happen when `au.baseAccounts` has older data than DB if round advances between `au.baseAccounts` and accountext reading. Retrying should sufficient - on next iteration we either do not hit this branch or data becomes synchronized.

File: ledger/acctupdates.go (hunk 16 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1902]: yes, I agree it might be a hard thing to test.

File: ledger/acctupdates.go (hunk 17 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1884]: I believe that if you just want to check if the error is a specific error, a type assertion would suffice. But using `errors.As` would not be incorrect.

**Review by tsachiherman — APPROVED** [review]
